python/ql/src/semmle/python/frameworks/Rsa.qll

```diff
@@ -0,0 +1,141 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `rsa` PyPI package.
+ * See https://stuvel.eu/python-rsa-doc/.
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.dataflow.new.TaintTracking
+private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
+
+/**
+ * Provides models for the `rsa` PyPI package.
+ * See https://stuvel.eu/python-rsa-doc/.
+ */
+private module Rsa {
+  /**
+   * A call to `rsa.newkeys`
+   *
+   * See https://stuvel.eu/python-rsa-doc/reference.html#rsa.newkeys
+   */
+  class RsaNewkeysCall extends Cryptography::PublicKey::KeyGeneration::RsaRange,
+    DataFlow::CallCfgNode {
+    RsaNewkeysCall() { this = API::moduleImport("rsa").getMember("newkeys").getACall() }
+
+    override DataFlow::Node getKeySizeArg() {
+      result in [this.getArg(0), this.getArgByName("nbits")]
+    }
+  }
+
+  /**
+   * A call to `rsa.encrypt`
+   *
+   * See https://stuvel.eu/python-rsa-doc/reference.html#rsa.encrypt
+   */
+  class RsaEncryptCall extends Cryptography::CryptographicOperation::Range, DataFlow::CallCfgNode {
+    RsaEncryptCall() { this = API::moduleImport("rsa").getMember("encrypt").getACall() }
+
+    override Cryptography::CryptographicAlgorithm getAlgorithm() { result.getName() = "RSA" }
+
+    override DataFlow::Node getAnInput() {
+      result in [this.getArg(0), this.getArgByName("message")]
+    }
+  }
+
+  /**
+   * A call to `rsa.decrypt`
+   *
+   * See https://stuvel.eu/python-rsa-doc/reference.html#rsa.decrypt
+   */
+  class RsaDecryptCall extends Cryptography::CryptographicOperation::Range, DataFlow::CallCfgNode {
+    RsaDecryptCall() { this = API::moduleImport("rsa").getMember("decrypt").getACall() }
+
+    override Cryptography::CryptographicAlgorithm getAlgorithm() { result.getName() = "RSA" }
+
+    override DataFlow::Node getAnInput() { result in [this.getArg(0), this.getArgByName("crypto")] }
+  }
+
+  /**
+   * A call to `rsa.sign`, which both hashes and signs in the input message.
+   *
+   * See https://stuvel.eu/python-rsa-doc/reference.html#rsa.sign
+   */
+  class RsaSignCall extends Cryptography::CryptographicOperation::Range, DataFlow::CallCfgNode {
+    RsaSignCall() { this = API::moduleImport("rsa").getMember("sign").getACall() }
+
+    override Cryptography::CryptographicAlgorithm getAlgorithm() {
+      // signature part
+      result.getName() = "RSA"
+      or
+      // hashing part
+      exists(StrConst str, DataFlow::Node hashNameArg |
+        hashNameArg in [this.getArg(2), this.getArgByName("hash_method")] and
+        DataFlow::exprNode(str).(DataFlow::LocalSourceNode).flowsTo(hashNameArg) and
```

Comment:
```suggestion
        DataFlow::exprNode(str) = hashNameArg.getALocalSource() and
```
I slightly prefer this form, as subtly implied in [the documentation](https://codeql.github.com/docs/codeql-language-guides/analyzing-data-flow-in-python/#examples).

Reply:
Yeah I think taus recently did these rewrites as well. I obviously just copied from the old code :smile: